bundles/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java

```diff
@@ -1105,10 +1118,21 @@ private static void internalPersist(Item item, TimeSeries timeSeries, @Nullable
         Iterator<HistoricItem> it = result.iterator();
         HistoricItem minimumHistoricItem = null;
 
+        Item baseItem = item;
+        if (baseItem instanceof GroupItem groupItem) {
+            baseItem = groupItem.getBaseItem();
+        }
+        Unit<?> unit = (baseItem instanceof NumberItem numberItem) ? numberItem.getUnit() : null;
+
         DecimalType minimum = null;
         while (it.hasNext()) {
             HistoricItem historicItem = it.next();
-            DecimalType value = historicItem.getState().as(DecimalType.class);
+            State state = historicItem.getState();
+            if ((unit != null) && state instanceof QuantityType<?> qtState) {
+                qtState = qtState.toUnit(unit);
+                state = qtState != null ? qtState : state;
+            }
+            DecimalType value = state.as(DecimalType.class);
```

Comment:
My understanding is that you get the unit from the item (or in case of a group from the base item). You then try to apply this unit to the state returned from the persistence service to ensure that all values have the correct unit. That is pretty logical and IMO correct. 

My problem is the next line (l. 1133): `.toUnit` returns `null` when a conversion is not possible. For example the persistence service returned a value of `5 m` but the item's unit is `s`. If that is the case, you fall back to the original (not converted value) and later extract the numeric only part of that state. IMO this is wrong, a value that can't be converted should be discarded instead. Maybe this is also a case where a warning should be logged, because that is for sure a severe configuration issue.

Reply:
@J-N-K You are of course right. I considered it but didn't think the chance of it happening was high (especially not with the current persistence services) and just avoided the null warning, but it is definitely better to skip and log a warning in the event that it would happen. I just inserted code for that.